code/books4u/api.py

```diff
@@ -16,3 +17,143 @@ def get_all_books(request):
+                break
+        if not exist:
+            category_o = BookCategory(name=category_)
+
+        exist = False
+        #check if the input author exists
+        authors = Author.objects.all()                        
+        for a in Author:
+            if a.name == author_:
+                exist = True
+                author_o = a
+                break
+        if not exist:
+            author_o = Author(name=author_, summary='')
+        
+        new_book = Book(ISBN=ISBN_, name=name_, publish_date=publish_date_, edition=edition_, category_o=category_, author_o=author_)
+        new_book.save()
+    
+    else:
+        return HttpResponse(json.dumps({'status': 'failure_method'}))
+
+""" call this method to check if the moderator should be notified that there is a pending new book
+    return a list of books needed to be verifid
+"""
+def get_pending_books():
+
+    pending_books = []
+    books = Book.objects.all()
+    for book in books:
+        if book.visibility == False:
+            pending_books.append(book)
+
+    return pending_books
+
+""" request contains two values:
+    a bool indicating permmited or not and a str indicating the ISBN of a pending book
+"""
+def commit_book(request):
+
+    if request.method == "POST":
+        try:
+            ISBN_ = request.POST.get('ISBN')
+            permmited = request.POST.get('permit')
+        except requests.exceptions.RequestException:
+            return HttpResponse(json.dumps({'status': 'missing_required_fields'}))
+
+        if permmited:
+            Book.objects.get(ISBN=ISBN_).visibility = True
+            return HttpResponse(json.dumps({'status': 'new book permmited'}))
+        else:
+            Book.objects.get(ISBN=ISBN_).delete()
+            return HttpResponse(json.dumps({'status': 'pending book rejected'}))
+
+    else:
+        return HttpResponse(json.dumps({'status': 'failure_method'}))
+
+def validate_rating(Rating):
+    if(Rating<=5&Rating>=1):
+        return True
+    else:
+        return False
+
+""" verify the user by checking the existence of email and corresponded password
+    request contains an email and a plain-text password
+"""
+def autentiate(request):
+    if request.method == "POST":
+        try:
+            email = request.POST.get('email')
+            password = request.POST.get('password')
+        except requests.exceptions.RequestException:
+            return HttpResponse(json.dumps({'status': 'missing_required_fields'}))
+
+        #check if the user exists
+        user = User.objects.all()                        
+        for u in User:
+            if u.email == email and check_password(password, u.password):
+                return HttpResponse(json.dumps({'status': 'user authorized'}))
+            
+        return HttpResponse(json.dumps({'status': 'incorrected email or password'}))
+
+    else:
+        return HttpResponse(json.dumps({'status': 'failure_method'}))
+
+def get_overallR_rating():
```

Comment:
- Why not merge 149, 152, 155 and 158into an api method which gets userinfo like how get_all_books() does. Besides, we don't need getters as helper methods to get attributes, since we can simply call getattr() or '.' for that purpose.
- 146 can remain in api if it is used like "
def get_overallR_rating(request):
    ...
    return HttpResponse(json.dumps(overall_rating), content_type="application/json")
"

Reply:
Yes. The only type of functions that are allowed to be in api.py are the ones that are linked in urls.py. If anyone defined a helper function, move it to utils.py